src/stores/events/EventsStore.ts

```diff
@@ -181,41 +185,115 @@ export default class EventsStore {
-		this.isLoadingRootEvents = true;
+			try {
+				const eventSource = this.api.sse.getEventSource({
+					type: 'event',
+					queryParams: {
+						startTimestamp: this.filterStore.eventsFilter.timestampFrom,
+						endTimestamp: this.filterStore.eventsFilter.timestampTo,
+						'name-values': this.filterStore.eventsFilter.names,
+						'type-values': this.filterStore.eventsFilter.eventTypes,
+					},
+				});
 
-		try {
-			const rootEventIds = await this.api.events.getEventTree(
-				this.filterStore.eventsFilter,
-				this.eventTreeAC.signal,
-			);
-			runInAction(() => {
-				this.eventTree = sortEventsByTimestamp(rootEventIds);
-			});
-			this.isLoadingRootEvents = false;
-		} catch (error) {
-			if (error.name !== 'AbortError') {
+				eventSource.onerror = e => {
+					if (eventSource.readyState !== EventSource.CONNECTING) {
+						console.error(e);
+						this.eventTree = [];
+						this.isExpandedMap.clear();
+						console.error('Error while loading root events', e);
+					}
+				};
+
+				eventSource.addEventListener('event', e => {
+					const event = JSON.parse((e as MessageEvent).data) as EventTreeNode;
+					this.addTreeNodeToEventTree(event);
+				});
+
+				eventSource.addEventListener('close', () => {
+					eventSource.close();
+					this.isLoadingRootEvents = false;
+					this.loadRootEvents();
+					this.eventsByParentIdMap = new Map();
+					res();
+				});
+			} catch (error) {
 				this.eventTreeStatusCode = error.status;
 				this.eventTree = [];
+				this.eventsByParentIdMap = new Map();
 				this.isExpandedMap.clear();
-				console.error('Error while loading root events', error);
 				this.isLoadingRootEvents = false;
+				console.error('Error while loading root events', error);
+				rej(error);
 			}
-		} finally {
-			this.eventTreeAC = null;
-		}
+		});
 	};
 
+	@action
+	private addTreeNodeToEventTree(node: EventTreeNode) {
+		const rootNode = this.eventTree.find(_rootNode => _rootNode.eventId === node.parentId);
+		if (rootNode) {
+			rootNode.childList.push(node);
+			return;
+		}
+
+		const accumulatedEvents = this.eventsByParentIdMap.get(node.eventId);
+		if (accumulatedEvents) {
+			node.childList.push(...accumulatedEvents);
+		}
+
+		if (node.parentId === 'null') {
+			this.eventTree = [...this.eventTree, node];
+			this.eventsByParentIdMap.delete(node.eventId);
+			return;
+		}
+
+		const accumulatedEventsByParentId = this.eventsByParentIdMap.get(node.parentId);
+
+		if (accumulatedEventsByParentId) {
+			accumulatedEventsByParentId.push(node);
+		} else {
+			this.eventsByParentIdMap.set(node.parentId, [node]);
+		}
+	}
+
+	@action
+	private async loadRootEvents() {
+		const loadEventById = async (eventId: string, childList: EventTreeNode[]) => {
+			try {
+				const event = await this.api.events.getEvent(eventId);
+				if (event.parentEventId === null) {
+					if (this.eventTree.find(node => node.eventId === event.eventId)) return;
+
+					const rootNode = convertEventActionToEventTreeNode(event);
```

Comment:
Maybe converting is unnecessary ? if we receive EventAction let's make it a tree of EventActions?